# Vehicle Insurance Fraud Analysis

# Project Context

This notebook documents the data understanding, preparation and exploratory analysis of the Vehicle Insurance Claim Fraud Detection dataset. The analysis focuses on identifying patterns associated with fraudulent and non-fraudulent insurance claims and assessing how the findings can support further investigation and claims prioritisation.

In [236]:
# Necessary dependencies/libraries

import pandas as pd

## 1. Load Dataset

Import the original Kaggle dataset into Python.

In [237]:
df = pd.read_csv("insurance_fraud.csv")

## 2. Understand Dataset

Examine the dataset structure, dimensions, attributes, data types and basic distributions.

---

### 2.1 Dataset Overview

In [238]:
df.head()

,Month,WeekOfMonth,DayOfWeek,Make,AccidentArea,DayOfWeekClaimed,MonthClaimed,WeekOfMonthClaimed,Sex,MaritalStatus,Age,Fault,PolicyType,VehicleCategory,VehiclePrice,FraudFound_P,PolicyNumber,RepNumber,Deductible,DriverRating,Days_Policy_Accident,Days_Policy_Claim,PastNumberOfClaims,AgeOfVehicle,AgeOfPolicyHolder,PoliceReportFiled,WitnessPresent,AgentType,NumberOfSuppliments,AddressChange_Claim,NumberOfCars,Year,BasePolicy
0,Dec,5,Wednesday,Honda,Urban,Tuesday,Jan,1,Female,Single,21,Policy Holder,Sport - Liability,Sport,more than 69000,0,1,12,300,1,more than 30,more than 30,none,3 years,26 to 30,No,No,External,none,1 year,3 to 4,1994,Liability
1,Jan,3,Wednesday,Honda,Urban,Monday,Jan,4,Male,Single,34,Policy Holder,Sport - Collision,Sport,more than 69000,0,2,15,400,4,more than 30,more than 30,none,6 years,31 to 35,Yes,No,External,none,no change,1 vehicle,1994,Collision
2,Oct,5,Friday,Honda,Urban,Thursday,Nov,2,Male,Married,47,Policy Holder,Sport - Collision,Sport,more than 69000,0,3,7,400,3,more than 30,more than 30,1,7 years,41 to 50,No,No,External,none,no change,1 vehicle,1994,Collision
3,Jun,2,Saturday,Toyota,Rural,Friday,Jul,1,Male,Married,65,Third Party,Sedan - Liability,Sport,20000 to 29000,0,4,4,400,2,more than 30,more than 30,1,more than 7,51 to 65,Yes,No,External,more than 5,no change,1 vehicle,1994,Liability
4,Jan,5,Monday,Honda,Urban,Tuesday,Feb,2,Female,Single,27,Third Party,Sport - Collision,Sport,more than 69000,0,5,3,400,1,more than 30,more than 30,none,5 years,31 to 35,No,No,External,none,no change,1 vehicle,1994,Collision


---

### 2.2 Dataset dimensions

In [239]:
df.shape

(15420, 33)

> 15, 420 insurance claim records, with 33 attributes per claim.

---

### 2.3 Dataset attributes

In [240]:
df.columns

Index(['Month', 'WeekOfMonth', 'DayOfWeek', 'Make', 'AccidentArea',
       'DayOfWeekClaimed', 'MonthClaimed', 'WeekOfMonthClaimed', 'Sex',
       'MaritalStatus', 'Age', 'Fault', 'PolicyType', 'VehicleCategory',
       'VehiclePrice', 'FraudFound_P', 'PolicyNumber', 'RepNumber',
       'Deductible', 'DriverRating', 'Days_Policy_Accident',
       'Days_Policy_Claim', 'PastNumberOfClaims', 'AgeOfVehicle',
       'AgeOfPolicyHolder', 'PoliceReportFiled', 'WitnessPresent', 'AgentType',
       'NumberOfSuppliments', 'AddressChange_Claim', 'NumberOfCars', 'Year',
       'BasePolicy'],
      dtype='str')

---

### 2.4 Dataset structure

In [241]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 15420 entries, 0 to 15419
Data columns (total 33 columns):
 #   Column                Non-Null Count  Dtype
---  ------                --------------  -----
 0   Month                 15420 non-null  str  
 1   WeekOfMonth           15420 non-null  int64
 2   DayOfWeek             15420 non-null  str  
 3   Make                  15420 non-null  str  
 4   AccidentArea          15420 non-null  str  
 5   DayOfWeekClaimed      15420 non-null  str  
 6   MonthClaimed          15420 non-null  str  
 7   WeekOfMonthClaimed    15420 non-null  int64
 8   Sex                   15420 non-null  str  
 9   MaritalStatus         15420 non-null  str  
 10  Age                   15420 non-null  int64
 11  Fault                 15420 non-null  str  
 12  PolicyType            15420 non-null  str  
 13  VehicleCategory       15420 non-null  str  
 14  VehiclePrice          15420 non-null  str  
 15  FraudFound_P          15420 non-null  int64
 16  PolicyNumber   

> No missing values: each column shows `15420 non-null`. Data Types: **9** Numerical and **24** String.

---

### 2.5 Inspect Target Variable, `FraudFound_P`

In [242]:
df["FraudFound_P"].value_counts()

FraudFound_P
0    14497
1      923
Name: count, dtype: int64

In [243]:
# Target Variable in percentages

df["FraudFound_P"].value_counts(normalize=True) * 100

FraudFound_P
0    94.014267
1     5.985733
Name: proportion, dtype: float64

_The target variable `FraudFound_P` is found to be imbalanced, with 94.01% of claims labelled 0 and 5.99% labelled 1._

## 3. Assess Data Quality

Identify missing values, duplicates, inconsistent values and other potential quality issues.

---

### 3.1 Unique values count for each (24) categorical attributes.

In [244]:
df.select_dtypes(include="str").nunique().sort_values()

AccidentArea             2
Sex                      2
Fault                    2
PoliceReportFiled        2
WitnessPresent           2
AgentType                2
BasePolicy               3
VehicleCategory          3
Days_Policy_Claim        4
PastNumberOfClaims       4
MaritalStatus            4
NumberOfSuppliments      4
AddressChange_Claim      5
NumberOfCars             5
Days_Policy_Accident     5
VehiclePrice             6
DayOfWeek                7
AgeOfVehicle             8
DayOfWeekClaimed         8
PolicyType               9
AgeOfPolicyHolder        9
Month                   12
MonthClaimed            13
Make                    19
dtype: int64

---

### 3.2 Inspecting the categorical values

In [245]:
df["Sex"].unique()

<StringArray>
['Female', 'Male']
Length: 2, dtype: str

In [246]:
for col in ["AccidentArea", "Fault", "PoliceReportFiled", "WitnessPresent", "AgentType"]:
  print(col, ":", df[col].unique(), "\n")

AccidentArea : <StringArray>
['Urban', 'Rural']
Length: 2, dtype: str 

Fault : <StringArray>
['Policy Holder', 'Third Party']
Length: 2, dtype: str 

PoliceReportFiled : <StringArray>
['No', 'Yes']
Length: 2, dtype: str 

WitnessPresent : <StringArray>
['No', 'Yes']
Length: 2, dtype: str 

AgentType : <StringArray>
['External', 'Internal']
Length: 2, dtype: str 



For the 6 binary variables `Sex`, `AccidentArea`, `Fault`, `PoliceReportFiled`, `WitnessPresent` and `AgentType`, we can conclude that:
* There are no spelling/capitalisation inconsistencies.
* The categories are consistent.
* There are no unexpected extra categories.

> Therefore, no need to clean or standardise these columns.

In [247]:
for col in df.select_dtypes(include="str").columns:
  print(col, ":", df[col].unique(), "\n")

Month : <StringArray>
['Dec', 'Jan', 'Oct', 'Jun', 'Feb', 'Nov', 'Apr', 'Mar', 'Aug', 'Jul', 'May',
 'Sep']
Length: 12, dtype: str 

DayOfWeek : <StringArray>
['Wednesday', 'Friday', 'Saturday', 'Monday', 'Tuesday', 'Sunday', 'Thursday']
Length: 7, dtype: str 

Make : <StringArray>
[    'Honda',    'Toyota',      'Ford',     'Mazda', 'Chevrolet',   'Pontiac',
    'Accura',     'Dodge',   'Mercury',    'Jaguar',    'Nisson',        'VW',
      'Saab',    'Saturn',    'Porche',       'BMW',   'Mecedes',   'Ferrari',
     'Lexus']
Length: 19, dtype: str 

AccidentArea : <StringArray>
['Urban', 'Rural']
Length: 2, dtype: str 

DayOfWeekClaimed : <StringArray>
[  'Tuesday',    'Monday',  'Thursday',    'Friday', 'Wednesday',  'Saturday',
    'Sunday',         '0']
Length: 8, dtype: str 

MonthClaimed : <StringArray>
['Jan', 'Nov', 'Jul', 'Feb', 'Mar', 'Dec', 'Apr', 'Aug', 'May', 'Jun', 'Sep',
 'Oct',   '0']
Length: 13, dtype: str 

Sex : <StringArray>
['Female', 'Male']
Length: 2, dtype: st

Attributes `DayOfWeekClaimed` and `MonthClaimed` have records with the numerical value **0**. We need to investigate their legitimacy and how they affects the analysis.

---

### 3.3 Identifying the 'errorneous' values

In [248]:
df["DayOfWeekClaimed"].value_counts()

DayOfWeekClaimed
Monday       3757
Tuesday      3375
Wednesday    2951
Thursday     2660
Friday       2497
Saturday      127
Sunday         52
0               1
Name: count, dtype: int64

In [249]:
df["MonthClaimed"].value_counts()

MonthClaimed
Jan    1446
May    1411
Mar    1348
Oct    1339
Jun    1293
Feb    1287
Nov    1285
Apr    1271
Sep    1242
Jul    1225
Dec    1146
Aug    1126
0         1
Name: count, dtype: int64

In [250]:
df[(df["DayOfWeekClaimed"] == "0") | (df["MonthClaimed"] == "0")]

,Month,WeekOfMonth,DayOfWeek,Make,AccidentArea,DayOfWeekClaimed,MonthClaimed,WeekOfMonthClaimed,Sex,MaritalStatus,Age,Fault,PolicyType,VehicleCategory,VehiclePrice,FraudFound_P,PolicyNumber,RepNumber,Deductible,DriverRating,Days_Policy_Accident,Days_Policy_Claim,PastNumberOfClaims,AgeOfVehicle,AgeOfPolicyHolder,PoliceReportFiled,WitnessPresent,AgentType,NumberOfSuppliments,AddressChange_Claim,NumberOfCars,Year,BasePolicy
1516,Jul,2,Monday,Honda,Rural,0,0,1,Male,Single,0,Policy Holder,Sedan - All Perils,Sedan,more than 69000,0,1517,15,400,2,more than 30,none,none,new,16 to 17,No,No,External,none,no change,1 vehicle,1994,All Perils


We can see that a single record at _index 1516_ has invalid values for both `DayOfWeekClaimed` and `MonthClaimed`. The decision is to keep the records since other attributes do not appear to be corrupted.

---

### 3.4 Inspecting Duplicates

In [251]:
df.duplicated().sum()

np.int64(0)

> **No** exact duplicate rows were found.

---

### 3.5 Inspecting Numerical Values

In [252]:
df.describe()

,WeekOfMonth,WeekOfMonthClaimed,Age,FraudFound_P,PolicyNumber,RepNumber,Deductible,DriverRating,Year
count,15420.000000,15420.000000,15420.000000,15420.000000,15420.000000,15420.000000,15420.000000,15420.000000,15420.000000
mean,2.788586,2.693969,39.855707,0.059857,7710.500000,8.483268,407.704280,2.487808,1994.866472
std,1.287585,1.259115,13.492377,0.237230,4451.514911,4.599948,43.950998,1.119453,0.803313
min,1.000000,1.000000,0.000000,0.000000,1.000000,1.000000,300.000000,1.000000,1994.000000
25%,2.000000,2.000000,31.000000,0.000000,3855.750000,5.000000,400.000000,1.000000,1994.000000
50%,3.000000,3.000000,38.000000,0.000000,7710.500000,8.000000,400.000000,2.000000,1995.000000
75%,4.000000,4.000000,48.000000,0.000000,11565.250000,12.000000,400.000000,3.000000,1996.000000
max,5.000000,5.000000,80.000000,1.000000,15420.000000,16.000000,700.000000,4.000000,1996.000000


`PolicyNumber` data suggests that this field is being used as an identifier for records. However, no definite conclusion is made at this stage.

`Age` column has a minimum value of **0**? This suggests further investigation.

In [253]:
(df["Age"] == 0).sum()

np.int64(320)

**320** records of `Age` have the value **0**!

In [254]:
df.loc[df["Age"] == 0, "AgeOfPolicyHolder"].value_counts()

AgeOfPolicyHolder
16 to 17    320
Name: count, dtype: int64

It would seem that `Age = 0` belongs to the category of `AgeOfPolicyHolder: 16 to 17`. But why?

In [255]:
df["Age"].value_counts().sort_index()

Age
0     320
16      9
17      6
18     48
19     32
     ... 
76     42
77     29
78     35
79     20
80     32
Name: count, Length: 66, dtype: int64

> `Age = 0` seems to be unusual encoded value that occurs for each of the 320 records whose `AgeOfPolicyHolder` is 16 to 17. Further investigation may be required.

---

### 3.6 Check for missing values explicitly

In [256]:
df.isna().sum().sum()

np.int64(0)

Section `2.4` demonstrated that we had **no** missing values and _np.int64(0)_ confirms it.

---

## 4. Clean the Data

We assess the dataset, identify the issues that are genuinely present, and apply preparation only where justified.

In [257]:
# Work on a clean dataframe.

df_clean = df.copy()

---

### 4.1 Issue 1: Handle the invalid `month` and `day of week` values

In [258]:
df_clean["DayOfWeekClaimed"] = df_clean["DayOfWeekClaimed"].replace("0", pd.NA)
df_clean["MonthClaimed"] = df_clean["MonthClaimed"].replace("0", pd.NA)

Now, **0** as both an invalid _Day_ and _Month_, has been replaced to a **Null** value.

---

### 4.2 Issue 2.1: Investigate `Age = 0`

In [259]:
df_clean.groupby("AgeOfPolicyHolder")["Age"].agg(["min", "max", "mean", "count"])

,min,max,mean,count
AgeOfPolicyHolder,,,,
16 to 17,0,0,0.000000,320
18 to 20,16,17,16.400000,15
21 to 25,18,20,18.814815,108
26 to 30,21,25,22.941272,613
31 to 35,26,35,30.548006,5593
36 to 40,36,45,40.483304,4043
41 to 50,46,55,50.423267,2828
51 to 65,56,65,60.441092,1392
over 65,66,80,72.783465,508


We investigated the relationship between `Age` and `AgeOfPolicyHolder` to try understanding what `Age = 0` meant. **However**, we can see that the `AgeOfPolicyHolder` seems to have been shifted upwards and we now need to understand this first.

---

### 4.3 Issue 2.2: Investigate `Age` and `AgeOfPolicyHolder`

We investigate the relationship between `Age` and `AgeOfPolicyHolder` to try understanding what `Age = 0` means.

In [260]:
pd.crosstab(df_clean["AgeOfPolicyHolder"], df_clean["Age"])

Age,0,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80
AgeOfPolicyHolder,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
16 to 17,320,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
18 to 20,0,9,6,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
21 to 25,0,0,0,48,32,28,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
26 to 30,0,0,0,0,0,0,127,125,122,135,104,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
31 to 35,0,0,0,0,0,0,0,0,0,0,0,535,540,560,552,596,550,544,574,573,569,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
36 to 40,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,406,410,384,435,383,423,401,404,411,386,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
41 to 50,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,296,308,291,265,290,279,276,253,288,282,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
51 to 65,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,146,144,134,138,156,145,112,136,146,135,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
over 65,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,42,31,32,32,27,40,45,32,35,34,42,29,35,20,32


The above analysis further confirms that `AgeOfPolicyHolder` have been wrongly inputed. Hence, we will create a new column named `AgeGroup` which contains the correct categories based on `Age`. 

The **320 records** with `Age = 0` were treated as having an unknown age. Since there was insufficient evidence to determine their actual age group, they were not assigned to an existing age category. Instead, they were assigned an `Unknown` category.

In [261]:
bins = [15, 17, 20, 25, 30, 35, 40, 50, 65, 80]

labels = [
  "16 to 17",
  "18 to 20",
  "21 to 25",
  "26 to 30",
  "31 to 35",
  "36 to 40",
  "41 to 50",
  "51 to 65",
  "over 65"
]

df_clean["AgeGroup"] = pd.cut(
  df_clean["Age"],
  bins=bins,
  labels=labels
)

df_clean["AgeGroup"] = df_clean["AgeGroup"].cat.add_categories("Unknown")

df_clean.loc[df_clean["Age"] == 0, "AgeGroup"] = "Unknown"

We now verify the new AgeGroup / Age relationship:

In [262]:
pd.crosstab(
  df_clean["AgeGroup"],
  df_clean["Age"],
  dropna=False
)

Age,0,16,17,18,19,20,21,22,23,24,25,26,27,28,29,30,31,32,33,34,35,36,37,38,39,40,41,42,43,44,45,46,47,48,49,50,51,52,53,54,55,56,57,58,59,60,61,62,63,64,65,66,67,68,69,70,71,72,73,74,75,76,77,78,79,80
AgeGroup,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,,
16 to 17,0,9,6,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
18 to 20,0,0,0,48,32,28,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
21 to 25,0,0,0,0,0,0,127,125,122,135,104,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
26 to 30,0,0,0,0,0,0,0,0,0,0,0,535,540,560,552,596,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
31 to 35,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,550,544,574,573,569,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
36 to 40,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,406,410,384,435,383,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
41 to 50,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,423,401,404,411,386,296,308,291,265,290,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
51 to 65,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,279,276,253,288,282,146,144,134,138,156,145,112,136,146,135,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
over 65,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,42,31,32,32,27,40,45,32,35,34,42,29,35,20,32


---

### 4.4 Issue 3: Investigate `PolicyNumber` and `RepNumber`

First, we determine uniqueness of `PolicyNumber`

In [263]:
df_clean["PolicyNumber"].nunique()

15420

Since we obtained **15420**, we can conclude that every row has a **unique** policy number. Hence the atribute is being used as an **identifier** and will potentially be discarded.

Then, we determine uniqueness of `RepNumber`

In [264]:
df_clean["RepNumber"].nunique()

16

**16** cannot mean that each record is unique. In fact, it identifies each **insurance representative** using a unique number (1 - 16) and a representative/agent may be associated to **multiple claims**.

---

## 5. Prepare Data Types and Attributes

Convert variables into appropriate data types and distinguish between numerical, categorical and identifier attributes.

In [265]:
df_clean.dtypes

Month                        str
WeekOfMonth                int64
DayOfWeek                    str
Make                         str
AccidentArea                 str
DayOfWeekClaimed             str
MonthClaimed                 str
WeekOfMonthClaimed         int64
Sex                          str
MaritalStatus                str
Age                        int64
Fault                        str
PolicyType                   str
VehicleCategory              str
VehiclePrice                 str
FraudFound_P               int64
PolicyNumber               int64
RepNumber                  int64
Deductible                 int64
DriverRating               int64
Days_Policy_Accident         str
Days_Policy_Claim            str
PastNumberOfClaims           str
AgeOfVehicle                 str
AgeOfPolicyHolder            str
PoliceReportFiled            str
WitnessPresent               str
AgentType                    str
NumberOfSuppliments          str
AddressChange_Claim          str
NumberOfCa

The attribute types were reviewed and although many are stored as integers or strings in python, their analytical interpretation was considered before deciding whether conversion was necessary.

At this stage, no conversion was deemed necesary. Categorical data were retained and their representation justified. However, depending on the analytical techniques or statistical modelling, appropriate conversions may be required.

The resulting attribute classification is summarised below.

| Attribute type  | Attributes            |
| --------------- | ----------------------------------------------------------------------------------------------------------------------------- |
| **Numerical**   | `Age`, `Deductible`   |
| **Categorical** | `Make`, `AccidentArea`, `Sex`, `MaritalStatus`, `Fault`, `PolicyType`, `VehicleCategory`, `VehiclePrice`, `DriverRating`, `Days_Policy_Accident`, `Days_Policy_Claim`, `PastNumberOfClaims`, `AgeOfVehicle`, `AgeOfPolicyHolder`, `PoliceReportFiled`, `WitnessPresent`, `AgentType`, `NumberOfSuppliments`, `AddressChange_Claim`, `NumberOfCars`, `BasePolicy`, `AgeGroup` |
| **Temporal**    | `Month`, `WeekOfMonth`, `DayOfWeek`, `DayOfWeekClaimed`, `MonthClaimed`, `WeekOfMonthClaimed`, `Year`      |
| **Identifier**  | `PolicyNumber`        |
| **Categorical Identifier** | `RepNumber`|
| **Target**      | `FraudFound_P`        |


---

## 6. Prepare Categorical and Numerical Data

Apply transformations or encoding where required by the selected analytical technique.

In [266]:
df_prepared = df_clean.copy()

### 6.1 Binary Categorical Variables

Explicit 0/1 Mapping for each attributes.

In [267]:
binary_cols = [
  "AccidentArea",
  "Sex",
  "Fault",
  "PoliceReportFiled",
  "WitnessPresent",
  "AgentType"
]

# ---------------------------------------------------

binary_mapping = {
  "AccidentArea": {"Rural": 0, "Urban": 1},
  "Sex": {"Female": 0, "Male": 1},
  "Fault": {"Third Party": 0, "Policy Holder": 1},
  "PoliceReportFiled": {"No": 0, "Yes": 1},
  "WitnessPresent": {"No": 0, "Yes": 1},
  "AgentType": {"External": 0, "Internal": 1}
}

for col, mapping in binary_mapping.items():
  df_prepared[col] = df_prepared[col].map(mapping)

# ---------------------------------------------------

df_prepared[binary_cols].head()

,AccidentArea,Sex,Fault,PoliceReportFiled,WitnessPresent,AgentType
0,1,0,1,0,0,0
1,1,1,1,1,0,0
2,1,1,1,0,0,0
3,0,1,0,1,0,0
4,1,0,0,0,0,0


### 6.2 Other Categorical Variables

Apply 1-Hot Encoding to remaining categorical attributes.

In [268]:
onehot_cols = [
  "Make",
  "MaritalStatus",
  "PolicyType",
  "VehicleCategory",
  "VehiclePrice",
  "DriverRating",
  "Days_Policy_Accident",
  "Days_Policy_Claim",
  "PastNumberOfClaims",
  "AgeOfVehicle",
  "NumberOfSuppliments",
  "AddressChange_Claim",
  "NumberOfCars",
  "BasePolicy",
  "AgeGroup",
  "Month",
  "WeekOfMonth",
  "DayOfWeek",
  "DayOfWeekClaimed",
  "MonthClaimed",
  "WeekOfMonthClaimed",
  "Year"
]

print(f"Number of variables to one-hot encode: {len(onehot_cols)}")
print(onehot_cols)

# ---------------------------------------------------

df_prepared = pd.get_dummies(
  df_prepared,
  columns=onehot_cols,
  dtype=int
)

# ---------------------------------------------------

df_prepared.shape

print(f"\nVariables after one-hot encoding: ")

# To show full column head/names
pd.set_option("display.max_columns", None)
print(df_prepared.columns.tolist())

df_prepared.head()

Number of variables to one-hot encode: 22
['Make', 'MaritalStatus', 'PolicyType', 'VehicleCategory', 'VehiclePrice', 'DriverRating', 'Days_Policy_Accident', 'Days_Policy_Claim', 'PastNumberOfClaims', 'AgeOfVehicle', 'NumberOfSuppliments', 'AddressChange_Claim', 'NumberOfCars', 'BasePolicy', 'AgeGroup', 'Month', 'WeekOfMonth', 'DayOfWeek', 'DayOfWeekClaimed', 'MonthClaimed', 'WeekOfMonthClaimed', 'Year']

Variables after one-hot encoding: 
['AccidentArea', 'Sex', 'Age', 'Fault', 'FraudFound_P', 'PolicyNumber', 'RepNumber', 'Deductible', 'AgeOfPolicyHolder', 'PoliceReportFiled', 'WitnessPresent', 'AgentType', 'Make_Accura', 'Make_BMW', 'Make_Chevrolet', 'Make_Dodge', 'Make_Ferrari', 'Make_Ford', 'Make_Honda', 'Make_Jaguar', 'Make_Lexus', 'Make_Mazda', 'Make_Mecedes', 'Make_Mercury', 'Make_Nisson', 'Make_Pontiac', 'Make_Porche', 'Make_Saab', 'Make_Saturn', 'Make_Toyota', 'Make_VW', 'MaritalStatus_Divorced', 'MaritalStatus_Married', 'MaritalStatus_Single', 'MaritalStatus_Widow', 'PolicyTyp

,AccidentArea,Sex,Age,Fault,FraudFound_P,PolicyNumber,RepNumber,Deductible,AgeOfPolicyHolder,PoliceReportFiled,WitnessPresent,AgentType,Make_Accura,Make_BMW,Make_Chevrolet,Make_Dodge,Make_Ferrari,Make_Ford,Make_Honda,Make_Jaguar,Make_Lexus,Make_Mazda,Make_Mecedes,Make_Mercury,Make_Nisson,Make_Pontiac,Make_Porche,Make_Saab,Make_Saturn,Make_Toyota,Make_VW,MaritalStatus_Divorced,MaritalStatus_Married,MaritalStatus_Single,MaritalStatus_Widow,PolicyType_Sedan - All Perils,PolicyType_Sedan - Collision,PolicyType_Sedan - Liability,PolicyType_Sport - All Perils,PolicyType_Sport - Collision,PolicyType_Sport - Liability,PolicyType_Utility - All Perils,PolicyType_Utility - Collision,PolicyType_Utility - Liability,VehicleCategory_Sedan,VehicleCategory_Sport,VehicleCategory_Utility,VehiclePrice_20000 to 29000,VehiclePrice_30000 to 39000,VehiclePrice_40000 to 59000,VehiclePrice_60000 to 69000,VehiclePrice_less than 20000,VehiclePrice_more than 69000,DriverRating_1,DriverRating_2,DriverRating_3,DriverRating_4,Days_Policy_Accident_1 to 7,Days_Policy_Accident_15 to 30,Days_Policy_Accident_8 to 15,Days_Policy_Accident_more than 30,Days_Policy_Accident_none,Days_Policy_Claim_15 to 30,Days_Policy_Claim_8 to 15,Days_Policy_Claim_more than 30,Days_Policy_Claim_none,PastNumberOfClaims_1,PastNumberOfClaims_2 to 4,PastNumberOfClaims_more than 4,PastNumberOfClaims_none,AgeOfVehicle_2 years,AgeOfVehicle_3 years,AgeOfVehicle_4 years,AgeOfVehicle_5 years,AgeOfVehicle_6 years,AgeOfVehicle_7 years,AgeOfVehicle_more than 7,AgeOfVehicle_new,NumberOfSuppliments_1 to 2,NumberOfSuppliments_3 to 5,NumberOfSuppliments_more than 5,NumberOfSuppliments_none,AddressChange_Claim_1 year,AddressChange_Claim_2 to 3 years,AddressChange_Claim_4 to 8 years,AddressChange_Claim_no change,AddressChange_Claim_under 6 months,NumberOfCars_1 vehicle,NumberOfCars_2 vehicles,NumberOfCars_3 to 4,NumberOfCars_5 to 8,NumberOfCars_more than 8,BasePolicy_All Perils,BasePolicy_Collision,BasePolicy_Liability,AgeGroup_16 to 17,AgeGroup_18 to 20,AgeGroup_21 to 25,AgeGroup_26 to 30,AgeGroup_31 to 35,AgeGroup_36 to 40,AgeGroup_41 to 50,AgeGroup_51 to 65,AgeGroup_over 65,AgeGroup_Unknown,Month_Apr,Month_Aug,Month_Dec,Month_Feb,Month_Jan,Month_Jul,Month_Jun,Month_Mar,Month_May,Month_Nov,Month_Oct,Month_Sep,WeekOfMonth_1,WeekOfMonth_2,WeekOfMonth_3,WeekOfMonth_4,WeekOfMonth_5,DayOfWeek_Friday,DayOfWeek_Monday,DayOfWeek_Saturday,DayOfWeek_Sunday,DayOfWeek_Thursday,DayOfWeek_Tuesday,DayOfWeek_Wednesday,DayOfWeekClaimed_Friday,DayOfWeekClaimed_Monday,DayOfWeekClaimed_Saturday,DayOfWeekClaimed_Sunday,DayOfWeekClaimed_Thursday,DayOfWeekClaimed_Tuesday,DayOfWeekClaimed_Wednesday,MonthClaimed_Apr,MonthClaimed_Aug,MonthClaimed_Dec,MonthClaimed_Feb,MonthClaimed_Jan,MonthClaimed_Jul,MonthClaimed_Jun,MonthClaimed_Mar,MonthClaimed_May,MonthClaimed_Nov,MonthClaimed_Oct,MonthClaimed_Sep,WeekOfMonthClaimed_1,WeekOfMonthClaimed_2,WeekOfMonthClaimed_3,WeekOfMonthClaimed_4,WeekOfMonthClaimed_5,Year_1994,Year_1995,Year_1996
0,1,0,21,1,0,1,12,300,26 to 30,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,0,1,1,0,0,0,0,0,0,1,0,0,0,1,0,0,0,0,1,0,1,0,0,0,0,0,0,0,0,0,1,1,0,0,0,0,0,0,1,0,0,0,0,1,0,0,1,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0
1,1,1,34,1,0,2,15,400,31 to 35,1,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0,0,1,0,0,0,1,0,0,0,1,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,0,1,0,0,0,1,0,1,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,1,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,1,0,0
2,1,1,47,1,0,3,7,400,41 to 50,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,1,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,1,0,1,0,0,0,0,0,1,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,1,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,1,0,0,0,1,0,0
3,0,1,65,0,0,4,4,400,51 to 65,1,0,0,0

---

## 7. Select Relevant Attributes

Identify variables that are useful for the business questions and analytical objectives while removing irrelevant or unsuitable attributes.

Removed:
- `PolicyNumber` since it is an **identifier**.
- `AgeOfPolicyHolder` since attribute is **inconsistent** with other attributes.
- `Age` since `AgeGroup` provides a more interpretable categorical representation for analysis and BI visualisation.
- `PolicyType` since it is a **composite attribute** formed from `VehicleCategory` and `BasePolicy`; its one-hot encoded columns were therefore removed.

In [ ]:
df_prepared = df_prepared.drop(
  columns=[
          "PolicyNumber", 
          "AgeOfPolicyHolder",
          "Age",
          "PolicyType_Sedan - All Perils",
          "PolicyType_Sedan - Collision",
          "PolicyType_Sedan - Liability",
          "PolicyType_Sport - All Perils",
          "PolicyType_Sport - Collision",
          "PolicyType_Sport - Liability",
          "PolicyType_Utility - All Perils",
          "PolicyType_Utility - Collision",
          "PolicyType_Utility - Liability"
  ]
)

In [270]:
# Current Columns in the dataset:

df_prepared.shape

print(f"\nColumns after selectng relevant attributes: ")

# To show full column head/names
pd.set_option("display.max_columns", None)
print(df_prepared.columns.tolist())

df_prepared.head()


Columns after selectng relevant attributes: 
['AccidentArea', 'Sex', 'Fault', 'FraudFound_P', 'RepNumber', 'Deductible', 'PoliceReportFiled', 'WitnessPresent', 'AgentType', 'Make_Accura', 'Make_BMW', 'Make_Chevrolet', 'Make_Dodge', 'Make_Ferrari', 'Make_Ford', 'Make_Honda', 'Make_Jaguar', 'Make_Lexus', 'Make_Mazda', 'Make_Mecedes', 'Make_Mercury', 'Make_Nisson', 'Make_Pontiac', 'Make_Porche', 'Make_Saab', 'Make_Saturn', 'Make_Toyota', 'Make_VW', 'MaritalStatus_Divorced', 'MaritalStatus_Married', 'MaritalStatus_Single', 'MaritalStatus_Widow', 'VehicleCategory_Sedan', 'VehicleCategory_Sport', 'VehicleCategory_Utility', 'VehiclePrice_20000 to 29000', 'VehiclePrice_30000 to 39000', 'VehiclePrice_40000 to 59000', 'VehiclePrice_60000 to 69000', 'VehiclePrice_less than 20000', 'VehiclePrice_more than 69000', 'DriverRating_1', 'DriverRating_2', 'DriverRating_3', 'DriverRating_4', 'Days_Policy_Accident_1 to 7', 'Days_Policy_Accident_15 to 30', 'Days_Policy_Accident_8 to 15', 'Days_Policy_Accid

,AccidentArea,Sex,Fault,FraudFound_P,RepNumber,Deductible,PoliceReportFiled,WitnessPresent,AgentType,Make_Accura,Make_BMW,Make_Chevrolet,Make_Dodge,Make_Ferrari,Make_Ford,Make_Honda,Make_Jaguar,Make_Lexus,Make_Mazda,Make_Mecedes,Make_Mercury,Make_Nisson,Make_Pontiac,Make_Porche,Make_Saab,Make_Saturn,Make_Toyota,Make_VW,MaritalStatus_Divorced,MaritalStatus_Married,MaritalStatus_Single,MaritalStatus_Widow,VehicleCategory_Sedan,VehicleCategory_Sport,VehicleCategory_Utility,VehiclePrice_20000 to 29000,VehiclePrice_30000 to 39000,VehiclePrice_40000 to 59000,VehiclePrice_60000 to 69000,VehiclePrice_less than 20000,VehiclePrice_more than 69000,DriverRating_1,DriverRating_2,DriverRating_3,DriverRating_4,Days_Policy_Accident_1 to 7,Days_Policy_Accident_15 to 30,Days_Policy_Accident_8 to 15,Days_Policy_Accident_more than 30,Days_Policy_Accident_none,Days_Policy_Claim_15 to 30,Days_Policy_Claim_8 to 15,Days_Policy_Claim_more than 30,Days_Policy_Claim_none,PastNumberOfClaims_1,PastNumberOfClaims_2 to 4,PastNumberOfClaims_more than 4,PastNumberOfClaims_none,AgeOfVehicle_2 years,AgeOfVehicle_3 years,AgeOfVehicle_4 years,AgeOfVehicle_5 years,AgeOfVehicle_6 years,AgeOfVehicle_7 years,AgeOfVehicle_more than 7,AgeOfVehicle_new,NumberOfSuppliments_1 to 2,NumberOfSuppliments_3 to 5,NumberOfSuppliments_more than 5,NumberOfSuppliments_none,AddressChange_Claim_1 year,AddressChange_Claim_2 to 3 years,AddressChange_Claim_4 to 8 years,AddressChange_Claim_no change,AddressChange_Claim_under 6 months,NumberOfCars_1 vehicle,NumberOfCars_2 vehicles,NumberOfCars_3 to 4,NumberOfCars_5 to 8,NumberOfCars_more than 8,BasePolicy_All Perils,BasePolicy_Collision,BasePolicy_Liability,AgeGroup_16 to 17,AgeGroup_18 to 20,AgeGroup_21 to 25,AgeGroup_26 to 30,AgeGroup_31 to 35,AgeGroup_36 to 40,AgeGroup_41 to 50,AgeGroup_51 to 65,AgeGroup_over 65,AgeGroup_Unknown,Month_Apr,Month_Aug,Month_Dec,Month_Feb,Month_Jan,Month_Jul,Month_Jun,Month_Mar,Month_May,Month_Nov,Month_Oct,Month_Sep,WeekOfMonth_1,WeekOfMonth_2,WeekOfMonth_3,WeekOfMonth_4,WeekOfMonth_5,DayOfWeek_Friday,DayOfWeek_Monday,DayOfWeek_Saturday,DayOfWeek_Sunday,DayOfWeek_Thursday,DayOfWeek_Tuesday,DayOfWeek_Wednesday,DayOfWeekClaimed_Friday,DayOfWeekClaimed_Monday,DayOfWeekClaimed_Saturday,DayOfWeekClaimed_Sunday,DayOfWeekClaimed_Thursday,DayOfWeekClaimed_Tuesday,DayOfWeekClaimed_Wednesday,MonthClaimed_Apr,MonthClaimed_Aug,MonthClaimed_Dec,MonthClaimed_Feb,MonthClaimed_Jan,MonthClaimed_Jul,MonthClaimed_Jun,MonthClaimed_Mar,MonthClaimed_May,MonthClaimed_Nov,MonthClaimed_Oct,MonthClaimed_Sep,WeekOfMonthClaimed_1,WeekOfMonthClaimed_2,WeekOfMonthClaimed_3,WeekOfMonthClaimed_4,WeekOfMonthClaimed_5,Year_1994,Year_1995,Year_1996
0,1,0,1,0,12,300,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,1,0,0,0,0,0,0,1,1,0,0,0,0,0,0,1,0,0,0,1,0,0,0,0,1,0,1,0,0,0,0,0,0,0,0,0,1,1,0,0,0,0,0,0,1,0,0,0,0,1,0,0,1,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0
1,1,1,1,0,15,400,1,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,1,0,0,0,0,0,0,1,0,0,0,1,0,0,0,1,0,0,0,1,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,0,1,0,0,0,1,0,1,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,1,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,1,0,1,0,0
2,1,1,1,0,7,400,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,1,0,0,1,0,0,0,0,1,0,0,0,1,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,1,0,1,0,0,0,0,0,1,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,1,1,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,1,0,0,0,1,0,0
3,0,1,0,0,4,400,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,1,0,0,0,1,0,1,0,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,1,0,1,0,0,0,0,0,0,0,0,0,1,0,0,0,1,0,0,0,0,1,0,1,0,0,0,0,0,0,1,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0,0,0,1,0,0
4,1,0,0,0,3,400,0,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,0,0,0,0,0,0,1,0,0,1,0,0,0,0,0,0,1,1,0,0,0,0,0,0,1,0,0,0,1,0,0,0,0,1,0,0,0,1,0,0,0,0,0,0,0,1,0,

---

## 8. Create the Prepared Dataset

Produce a clean and analysis-ready dataset for subsequent analysis, modelling and BI development.

### 8.1 Final Dataset Validation

In [271]:
print("Shape:", df_prepared.shape)
print("Missing values:", df_prepared.isna().sum().sum())
print("Duplicate rows:", df_prepared.duplicated().sum())
print("Data types:")
print(df_prepared.dtypes.value_counts())

Shape: (15420, 144)
Missing values: 0
Duplicate rows: 0
Data types:
int64    144
Name: count, dtype: int64


### 8.2 Verify Target Variable

In [272]:
df_prepared["FraudFound_P"].value_counts()

FraudFound_P
0    14497
1      923
Name: count, dtype: int64

### 8.3 Save Prepared Dataset

In [273]:
df_prepared.to_csv("insurance_fraud_prepared.csv", index=False)

import os
print(os.path.exists("insurance_fraud_prepared.csv"))

True


---